File: backend/app/api/v1/endpoints/estimates.py
"""Estimate management endpoints."""

import secrets
import uuid
from datetime import datetime
from uuid import UUID

from fastapi import APIRouter, HTTPException, Query, status
from sqlalchemy import select, func
from sqlalchemy.orm import selectinload

from app.api.v1.deps import CurrentUser, CurrentOrganization, DBSession
from app.models.estimate import Estimate, EstimateLineItem, EstimateTemplate, EstimateStatus
from app.models.job import Job
from app.schemas.estimate import (
    EstimateCreate, EstimateUpdate, EstimateResponse,
    EstimateLineItemCreate, EstimateLineItemUpdate, EstimateLineItemResponse,
    EstimateTemplateCreate, EstimateTemplateResponse,
)

router = APIRouter()


def generate_estimate_number(org_id: UUID) -> str:
    """Generate a unique estimate number."""
    date_part = datetime.utcnow().strftime("%Y%m%d")
    random_part = uuid.uuid4().hex[:4].upper()
    return f"EST-{date_part}-{random_part}"


def calculate_estimate_totals(estimate: Estimate) -> None:
    """Calculate and update estimate totals."""
    subtotal = 0.0
    cost = 0.0
    good_total = 0.0
    better_total = 0.0
    best_total = 0.0
    
    for item in estimate.line_items:
        item.total = item.quantity * item.unit_price
        item.total_cost = item.quantity * item.unit_cost
        
        if item.is_selected:
            if item.option == "good":
                good_total += item.total
            elif item.option == "best":
                best_total += item.total
            else:  # better or no option
                subtotal += item.total
                cost += item.total_cost
                better_total += item.total
    
    # Apply discount
    if estimate.discount_percent > 0:
        discount = subtotal * (estimate.discount_percent / 100)
    else:
        discount = estimate.discount_amount
    
    subtotal_after_discount = subtotal - discount
    
    # Calculate tax
    tax_amount = subtotal_after_discount * (estimate.tax_rate / 100)
    
    # Update estimate
    estimate.subtotal = subtotal
    estimate.tax_amount = tax_amount
    estimate.total = subtotal_after_discount + tax_amount
    estimate.cost = cost
    
    # Good/Better/Best totals
    if estimate.presentation_type.value == "good_better_best":
        estimate.good_total = good_total + (good_total * estimate.tax_rate / 100)
        estimate.better_total = better_total + (better_total * estimate.tax_rate / 100)
        estimate.best_total = best_total + (best_total * estimate.tax_rate / 100)
    
    # Calculate monthly payment if financing enabled
    if estimate.include_financing and estimate.financing_term_months and estimate.financing_rate:
        # Simple calculation (not accounting for APR compounding)
        monthly_rate = estimate.financing_rate / 100 / 12
        months = estimate.financing_term_months
        if monthly_rate > 0:
            estimate.monthly_payment = estimate.total * (monthly_rate * (1 + monthly_rate) ** months) / ((1 + monthly_rate) ** months - 1)
        else:
            estimate.monthly_payment = estimate.total / months


# ============ Estimate Templates ============

@router.get("/templates", response_model=list[EstimateTemplateResponse])
async def list_templates(
    db: DBSession,
    current_user: CurrentUser,
    organization: CurrentOrganization,
    category: str | None = None,
):
    """List estimate templates."""
    query = (
        select(EstimateTemplate)
        .where(
            EstimateTemplate.organization_id == organization.id,
            EstimateTemplate.is_active == True,
        )
    )
    
    if category:
        query = query.where(EstimateTemplate.category == category)
    
    result = await db.execute(query.order_by(EstimateTemplate.name))
    return result.scalars().all()


@router.post("/templates", response_model=EstimateTemplateResponse, status_code=status.HTTP_201_CREATED)
async def create_template(
    data: EstimateTemplateCreate,
    db: DBSession,
    current_user: CurrentUser,
    organization: CurrentOrganization,
):
    """Create an estimate template."""
    template = EstimateTemplate(
        organization_id=organization.id,
        **data.model_dump(),
    )
    db.add(template)
    await db.commit()
    await db.refresh(template)
    return template


# ============ Estimates ============

@router.get("", response_model=list[EstimateResponse])
async def list_estimates(
    db: DBSession,
    current_user: CurrentUser,
    organization: CurrentOrganization,
    job_id: UUID | None = None,
    status: EstimateStatus | None = None,
):
    """List estimates."""
    # Get job IDs for this organization
    jobs_query = select(Job.id).where(Job.organization_id == organization.id)
    
    query = (
        select(Estimate)
        .options(selectinload(Estimate.line_items))
        .where(Estimate.job_id.in_(jobs_query))
    )
    
    if job_id:
        query = query.where(Estimate.job_id == job_id)
    
    if status:
        query = query.where(Estimate.status == status)
    
    result = await db.execute(query.order_by(Estimate.created_at.desc()))
    return result.scalars().all()


@router.post("", response_model=EstimateResponse, status_code=status.HTTP_201_CREATED)
async def create_estimate(
    data: EstimateCreate,
    db: DBSession,
    current_user: CurrentUser,
    organization: CurrentOrganization,
):
    """Create a new estimate."""
    # Verify job exists and belongs to organization
    result = await db.execute(
        select(Job).where(Job.id == data.job_id, Job.organization_id == organization.id)
    )
    if not result.scalar_one_or_none():
        raise HTTPException(status_code=404, detail="Job not found")
    
    # Create estimate
    estimate_data = data.model_dump(exclude={"line_items"})
    estimate = Estimate(
        estimate_number=generate_estimate_number(organization.id),
        share_token=secrets.token_urlsafe(32),
        **estimate_data,
    )
    db.add(estimate)
    await db.flush()
    
    # Create line items
    if data.line_items:
        for i, item_data in enumerate(data.line_items):
            item = EstimateLineItem(
                estimate_id=estimate.id,
                order=item_data.order or i,
                **item_data.model_dump(exclude={"order"}),
            )
            db.add(item)
    
    await db.flush()
    
    # Calculate totals
    result = await db.execute(
        select(Estimate)
        .options(selectinload(Estimate.line_items))
        .where(Estimate.id == estimate.id)
    )
    estimate = result.scalar_one()
    calculate_estimate_totals(estimate)
    
    await db.commit()
    await db.refresh(estimate)
    
    return estimate


@router.get("/{estimate_id}", response_model=EstimateResponse)
async def get_estimate(
    estimate_id: UUID,
    db: DBSession,
    current_user: CurrentUser,
    organization: CurrentOrganization,
):
    """Get a specific estimate."""
    # Get job IDs for this organization
    jobs_query = select(Job.id).where(Job.organization_id == organization.id)
    
    result = await db.execute(
        select(Estimate)
        .options(selectinload(Estimate.line_items))
        .where(Estimate.id == estimate_id, Estimate.job_id.in_(jobs_query))
    )
    estimate = result.scalar_one_or_none()
    
    if not estimate:
        raise HTTPException(status_code=404, detail="Estimate not found")
    
    return estimate


@router.patch("/{estimate_id}", response_model=EstimateResponse)
async def update_estimate(
    estimate_id: UUID,
    data: EstimateUpdate,
    db: DBSession,
    current_user: CurrentUser,
    organization: CurrentOrganization,
):
    """Update an estimate."""
    # Get job IDs for this organization
    jobs_query = select(Job.id).where(Job.organization_id == organization.id)
    
    result = await db.execute(
        select(Estimate)
        .options(selectinload(Estimate.line_items))
        .where(Estimate.id == estimate_id, Estimate.job_id.in_(jobs_query))
    )
    estimate = result.scalar_one_or_none()
    
    if not estimate:
        raise HTTPException(status_code=404, detail="Estimate not found")
    
    update_data = data.model_dump(exclude_unset=True)
    
    # Track status changes
    if "status" in update_data:
        new_status = update_data["status"]
        if new_status == EstimateStatus.SENT and estimate.status == EstimateStatus.DRAFT:
            update_data["sent_at"] = datetime.utcnow()
        elif new_status == EstimateStatus.APPROVED:
            update_data["approved_at"] = datetime.utcnow()
    
    for field, value in update_data.items():
        setattr(estimate, field, value)
    
    # Recalculate totals
    calculate_estimate_totals(estimate)
    
    await db.commit()
    await db.refresh(estimate)
    
    return estimate


@router.delete("/{estimate_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_estimate(
    estimate_id: UUID,
    db: DBSession,
    current_user: CurrentUser,
    organization: CurrentOrganization,
):
    """Delete an estimate."""
    # Get job IDs for this organization
    jobs_query = select(Job.id).where(Job.organization_id == organization.id)
    
    result = await db.execute(
        select(Estimate)
        .where(Estimate.id == estimate_id, Estimate.job_id.in_(jobs_query))
    )
    estimate = result.scalar_one_or_none()
    
    if not estimate:
        raise HTTPException(status_code=404, detail="Estimate not found")
    
    await db.delete(estimate)
    await db.commit()


# ============ Line Items ============

@router.post("/{estimate_id}/items", response_model=EstimateLineItemResponse, status_code=status.HTTP_201_CREATED)
async def add_line_item(
    estimate_id: UUID,
    data: EstimateLineItemCreate,
    db: DBSession,
    current_user: CurrentUser,
    organization: CurrentOrganization,
):
    """Add a line item to an estimate."""
    # Get job IDs for this organization
    jobs_query = select(Job.id).where(Job.organization_id == organization.id)
    
    result = await db.execute(
        select(Estimate)
        .options(selectinload(Estimate.line_items))
        .where(Estimate.id == estimate_id, Estimate.job_id.in_(jobs_query))
    )
    estimate = result.scalar_one_or_none()
    
    if not estimate:
        raise HTTPException(status_code=404, detail="Estimate not found")
    
    # Calculate total
    total = data.quantity * data.unit_price
    total_cost = data.quantity * data.unit_cost
    
    item = EstimateLineItem(
        estimate_id=estimate_id,
        total=total,
        total_cost=total_cost,
        **data.model_dump(),
    )
    db.add(item)
    await db.flush()
    
    # Recalculate estimate totals
    estimate.line_items.append(item)
    calculate_estimate_totals(estimate)
    
    await db.commit()
    await db.refresh(item)
    
    return item


@router.patch("/{estimate_id}/items/{item_id}", response_model=EstimateLineItemResponse)
async def update_line_item(
    estimate_id: UUID,
    item_id: UUID,
    data: EstimateLineItemUpdate,
    db: DBSession,
    current_user: CurrentUser,
    organization: CurrentOrganization,
):
    """Update a line item."""
    # Get job IDs for this organization
    jobs_query = select(Job.id).where(Job.organization_id == organization.id)
    
    result = await db.execute(
        select(Estimate)
        .options(selectinload(Estimate.line_items))
        .where(Estimate.id == estimate_id, Estimate.job_id.in_(jobs_query))
    )
    estimate = result.scalar_one_or_none()
    
    if not estimate:
        raise HTTPException(status_code=404, detail="Estimate not found")
    
    result = await db.execute(
        select(EstimateLineItem)
        .where(EstimateLineItem.id == item_id, EstimateLineItem.estimate_id == estimate_id)
    )
    item = result.scalar_one_or_none()
    
    if not item:
        raise HTTPException(status_code=404, detail="Line item not found")
    
    update_data = data.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(item, field, value)
    
    # Recalculate item total
    item.total = item.quantity * item.unit_price
    item.total_cost = item.quantity * item.unit_cost
    
    # Recalculate estimate totals
    calculate_estimate_totals(estimate)
    
    await db.commit()
    await db.refresh(item)
    
    return item


@router.delete("/{estimate_id}/items/{item_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_line_item(
    estimate_id: UUID,
    item_id: UUID,
    db: DBSession,
    current_user: CurrentUser,
    organization: CurrentOrganization,
):
    """Delete a line item."""
    # Get job IDs for this organization
    jobs_query = select(Job.id).where(Job.organization_id == organization.id)
    
    result = await db.execute(
        select(Estimate)
        .options(selectinload(Estimate.line_items))
        .where(Estimate.id == estimate_id, Estimate.job_id.in_(jobs_query))
    )
    estimate = result.scalar_one_or_none()
    
    if not estimate:
        raise HTTPException(status_code=404, detail="Estimate not found")
    
    result = await db.execute(
        select(EstimateLineItem)
        .where(EstimateLineItem.id == item_id, EstimateLineItem.estimate_id == estimate_id)
    )
    item = result.scalar_one_or_none()
    
    if not item:
        raise HTTPException(status_code=404, detail="Line item not found")
    
    await db.delete(item)
    
    # Recalculate estimate totals
    estimate.line_items = [i for i in estimate.line_items if i.id != item_id]
    calculate_estimate_totals(estimate)
    
    await db.commit()


# ============ Actions ============

@router.post("/{estimate_id}/send", response_model=EstimateResponse)
async def send_estimate(
    estimate_id: UUID,
    db: DBSession,
    current_user: CurrentUser,
    organization: CurrentOrganization,
):
    """Send estimate to customer."""
    # Get job IDs for this organization
    jobs_query = select(Job.id).where(Job.organization_id == organization.id)
    
    result = await db.execute(
        select(Estimate)
        .options(selectinload(Estimate.line_items))
        .where(Estimate.id == estimate_id, Estimate.job_id.in_(jobs_query))
    )
    estimate = result.scalar_one_or_none()
    
    if not estimate:
        raise HTTPException(status_code=404, detail="Estimate not found")
    
    estimate.status = EstimateStatus.SENT
    estimate.sent_at = datetime.utcnow()
    
    # TODO: Actually send email with proposal link
    # The share_token can be used to create a public link
    
    await db.commit()
    await db.refresh(estimate)
    
    return estimate

File: frontend/src/stores/estimates.ts
/**
 * Estimates Store - Estimate state management
 *
 * Manages estimates list, filtering, and CRUD operations.
 */

import { create } from 'zustand'
import { api } from '@/lib/api'

export type EstimateStatus = 'draft' | 'sent' | 'viewed' | 'accepted' | 'declined' | 'expired'

export interface EstimateLineItem {
  id: string
  description: string
  quantity: number
  unitPrice: number
  unitCost: number
  total: number
  totalCost: number
  option: 'good' | 'better' | 'best' | null
  isSelected: boolean
  sortOrder: number
}

export interface Estimate {
  id: string
  estimateNumber: string
  jobId: string | null
  contactId: string | null
  
  // Details
  name: string
  description: string | null
  
  // Status
  status: EstimateStatus
  sentAt: string | null
  viewedAt: string | null
  acceptedAt: string | null
  declinedAt: string | null
  expiresAt: string | null
  
  // Financial
  subtotal: number
  discountPercent: number
  discountAmount: number
  taxRate: number
  taxAmount: number
  total: number
  cost: number
  profitMargin: number
  
  // Line items
  lineItems: EstimateLineItem[]
  
  // Sharing
  shareToken: string | null
  shareUrl: string | null
  
  // Metadata
  createdAt: string
  updatedAt: string
}

export interface EstimateTemplate {
  id: string
  name: string
  description: string | null
  lineItems: Omit<EstimateLineItem, 'id'>[]
  isDefault: boolean
  createdAt: string
}

export interface EstimatesFilters {
  status?: EstimateStatus[]
  search?: string
  jobId?: string
  contactId?: string
}

export interface EstimatesState {
  // Data
  estimates: Estimate[]
  selectedEstimate: Estimate | null
  templates: EstimateTemplate[]

  // UI State
  isLoading: boolean
  error: string | null
  filters: EstimatesFilters

  // Pagination
  page: number
  pageSize: number
  totalCount: number
  totalPages: number

  // Actions
  fetchEstimates: () => Promise<void>
  fetchEstimate: (id: string) => Promise<void>
  fetchTemplates: () => Promise<void>
  createEstimate: (data: Partial<Estimate>) => Promise<Estimate>
  updateEstimate: (id: string, data: Partial<Estimate>) => Promise<void>
  deleteEstimate: (id: string) => Promise<void>
  sendEstimate: (id: string) => Promise<void>

  // Line item actions
  addLineItem: (estimateId: string, item: Partial<EstimateLineItem>) => Promise<void>
  updateLineItem: (estimateId: string, itemId: string, data: Partial<EstimateLineItem>) => Promise<void>
  deleteLineItem: (estimateId: string, itemId: string) => Promise<void>

  // Filter actions
  setFilters: (filters: Partial<EstimatesFilters>) => void
  clearFilters: () => void
  setPage: (page: number) => void

  // Selection
  setSelectedEstimate: (estimate: Estimate | null) => void
  clearError: () => void
}

const defaultFilters: EstimatesFilters = {
  status: undefined,
  search: undefined,
  jobId: undefined,
  contactId: undefined,
}

// Transform backend response to frontend format
function transformEstimate(data: any): Estimate {
  return {
    id: data.id,
    estimateNumber: data.estimate_number,
    jobId: data.job_id,
    contactId: data.contact_id,
    name: data.name,
    description: data.description,
    status: data.status,
    sentAt: data.sent_at,
    viewedAt: data.viewed_at,
    acceptedAt: data.accepted_at,
    declinedAt: data.declined_at,
    expiresAt: data.expires_at,
    subtotal: data.subtotal,
    discountPercent: data.discount_percent,
    discountAmount: data.discount_amount,
    taxRate: data.tax_rate,
    taxAmount: data.tax_amount,
    total: data.total,
    cost: data.cost,
    profitMargin: data.profit_margin,
    lineItems: (data.line_items || []).map(transformLineItem),
    shareToken: data.share_token,
    shareUrl: data.share_url,
    createdAt: data.created_at,
    updatedAt: data.updated_at,
  }
}

function transformLineItem(data: any): EstimateLineItem {
  return {
    id: data.id,
    description: data.description,
    quantity: data.quantity,
    unitPrice: data.unit_price,
    unitCost: data.unit_cost,
    total: data.total,
    totalCost: data.total_cost,
    option: data.option,
    isSelected: data.is_selected,
    sortOrder: data.sort_order,
  }
}

function transformTemplate(data: any): EstimateTemplate {
  return {
    id: data.id,
    name: data.name,
    description: data.description,
    lineItems: (data.line_items || []).map((item: any) => ({
      description: item.description,
      quantity: item.quantity,
      unitPrice: item.unit_price,
      unitCost: item.unit_cost,
      total: item.total,
      totalCost: item.total_cost,
      option: item.option,
      isSelected: item.is_selected,
      sortOrder: item.sort_order,
    })),
    isDefault: data.is_default,
    createdAt: data.created_at,
  }
}

export const useEstimatesStore = create<EstimatesState>()((set, get) => ({
  // Initial state
  estimates: [],
  selectedEstimate: null,
  templates: [],
  isLoading: false,
  error: null,
  filters: defaultFilters,
  page: 1,
  pageSize: 50,
  totalCount: 0,
  totalPages: 0,

  // Fetch estimates list
  fetchEstimates: async () => {
    set({ isLoading: true, error: null })
    try {
      const { filters, page, pageSize } = get()
      const params = new URLSearchParams()

      params.append('page', page.toString())
      params.append('page_size', pageSize.toString())

      if (filters.status?.length) {
        filters.status.forEach((s) => params.append('status', s))
      }
      if (filters.search) {
        params.append('search', filters.search)
      }
      if (filters.jobId) {
        params.append('job_id', filters.jobId)
      }
      if (filters.contactId) {
        params.append('contact_id', filters.contactId)
      }

      const response = await api.get(`/estimates?${params.toString()}`)
      const { items, total, pages } = response.data

      set({
        estimates: items.map(transformEstimate),
        totalCount: total,
        totalPages: pages,
        isLoading: false,
      })
    } catch (error: any) {
      set({
        error: error.response?.data?.detail || 'Failed to fetch estimates',
        isLoading: false,
      })
    }
  },

  // Fetch single estimate
  fetchEstimate: async (id: string) => {
    set({ isLoading: true, error: null })
    try {
      const response = await api.get(`/estimates/${id}`)
      set({ selectedEstimate: transformEstimate(response.data), isLoading: false })
    } catch (error: any) {
      set({
        error: error.response?.data?.detail || 'Failed to fetch estimate',
        isLoading: false,
      })
    }
  },

  // Fetch templates
  fetchTemplates: async () => {
    try {
      const response = await api.get('/estimates/templates')
      set({ templates: response.data.map(transformTemplate) })
    } catch (error: any) {
      console.error('Failed to fetch templates:', error)
    }
  },

  // Create estimate
  createEstimate: async (data: Partial<Estimate>) => {
    set({ isLoading: true, error: null })
    try {
      const payload = {
        name: data.name,
        description: data.description,
        job_id: data.jobId,
        contact_id: data.contactId,
      }
      const response = await api.post('/estimates', payload)
      const newEstimate = transformEstimate(response.data)

      set((state) => ({
        estimates: [newEstimate, ...state.estimates],
        isLoading: false,
      }))

      return newEstimate
    } catch (error: any) {
      set({
        error: error.response?.data?.detail || 'Failed to create estimate',
        isLoading: false,
      })
      throw error
    }
  },

  // Update estimate
  updateEstimate: async (id: string, data: Partial<Estimate>) => {
    set({ isLoading: true, error: null })
    try {
      const payload: Record<string, any> = {}
      if (data.name !== undefined) payload.name = data.name
      if (data.description !== undefined) payload.description = data.description
      if (data.discountPercent !== undefined) payload.discount_percent = data.discountPercent
      if (data.discountAmount !== undefined) payload.discount_amount = data.discountAmount
      if (data.taxRate !== undefined) payload.tax_rate = data.taxRate

      const response = await api.patch(`/estimates/${id}`, payload)
      const updated = transformEstimate(response.data)

      set((state) => ({
        estimates: state.estimates.map((e) => (e.id === id ? updated : e)),
        selectedEstimate: state.selectedEstimate?.id === id ? updated : state.selectedEstimate,
        isLoading: false,
      }))
    } catch (error: any) {
      set({
        error: error.response?.data?.detail || 'Failed to update estimate',
        isLoading: false,
      })
      throw error
    }
  },

  // Delete estimate
  deleteEstimate: async (id: string) => {
    set({ isLoading: true, error: null })
    try {
      await api.delete(`/estimates/${id}`)

      set((state) => ({
        estimates: state.estimates.filter((e) => e.id !== id),
        selectedEstimate: state.selectedEstimate?.id === id ? null : state.selectedEstimate,
        isLoading: false,
      }))
    } catch (error: any) {
      set({
        error: error.response?.data?.detail || 'Failed to delete estimate',
        isLoading: false,
      })
      throw error
    }
  },

  // Send estimate
  sendEstimate: async (id: string) => {
    set({ isLoading: true, error: null })
    try {
      const response = await api.post(`/estimates/${id}/send`)
      const updated = transformEstimate(response.data)

      set((state) => ({
        estimates: state.estimates.map((e) => (e.id === id ? updated : e)),
        selectedEstimate: state.selectedEstimate?.id === id ? updated : state.selectedEstimate,
        isLoading: false,
      }))
    } catch (error: any) {
      set({
        error: error.response?.data?.detail || 'Failed to send estimate',
        isLoading: false,
      })
      throw error
    }
  },

  // Add line item
  addLineItem: async (estimateId: string, item: Partial<EstimateLineItem>) => {
    try {
      const payload = {
        description: item.description,
        quantity: item.quantity || 1,
        unit_price: item.unitPrice || 0,
        unit_cost: item.unitCost || 0,
        option: item.option,
        is_selected: item.isSelected ?? true,
      }
      await api.post(`/estimates/${estimateId}/items`, payload)
      // Refresh the estimate to get updated totals
      await get().fetchEstimate(estimateId)
    } catch (error: any) {
      set({ error: error.response?.data?.detail || 'Failed to add line item' })
      throw error
    }
  },

  // Update line item
  updateLineItem: async (estimateId: string, itemId: string, data: Partial<EstimateLineItem>) => {
    try {
      const payload: Record<string, any> = {}
      if (data.description !== undefined) payload.description = data.description
      if (data.quantity !== undefined) payload.quantity = data.quantity
      if (data.unitPrice !== undefined) payload.unit_price = data.unitPrice
      if (data.unitCost !== undefined) payload.unit_cost = data.unitCost
      if (data.option !== undefined) payload.option = data.option
      if (data.isSelected !== undefined) payload.is_selected = data.isSelected

      await api.patch(`/estimates/${estimateId}/items/${itemId}`, payload)
      // Refresh the estimate to get updated totals
      await get().fetchEstimate(estimateId)
    } catch (error: any) {
      set({ error: error.response?.data?.detail || 'Failed to update line item' })
      throw error
    }
  },

  // Delete line item
  deleteLineItem: async (estimateId: string, itemId: string) => {
    try {
      await api.delete(`/estimates/${estimateId}/items/${itemId}`)
      // Refresh the estimate to get updated totals
      await get().fetchEstimate(estimateId)
    } catch (error: any) {
      set({ error: error.response?.data?.detail || 'Failed to delete line item' })
      throw error
    }
  },

  // Filter actions
  setFilters: (filters: Partial<EstimatesFilters>) => {
    set((state) => ({
      filters: { ...state.filters, ...filters },
      page: 1,
    }))
    get().fetchEstimates()
  },

  clearFilters: () => {
    set({ filters: defaultFilters, page: 1 })
    get().fetchEstimates()
  },

  setPage: (page) => {
    set({ page })
    get().fetchEstimates()
  },

  setSelectedEstimate: (estimate) => set({ selectedEstimate: estimate }),

  clearError: () => set({ error: null }),
}))
